### DISCUSSION TITLE
Can Tex.FromFile() safely be called from a different thread?

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#247] [Q&A]
I guess not as it needs to create a GPU resource, and I suspect that can only be done on the main thread

**Comment by maluoi** [comment]
Currently, it does need to be called from the main thread. This is something I hope to alleviate with the upcoming async asset feature, but it is currently a limitation! If you're on a separate thread, you can use `SK.ExecuteOnMain` to defer the action to the correct thread.

**Comment by paulmelis** [comment]
I guess the async texture loading is now done in the latest available preview. But what about `Model.FromFile()` and friends? I currently call it on a separate thread to off-load reading of large models (10-100s of MB), but although it doesn't crash it does lockup the rendering. I guess that's expected currently?

**Reply by maluoi** [reply]
That is not expected at the moment! Possibly some very brief pauses on OpenGL systems during GPU upload, but that should be the most of it! I did a pass to make sure all asset creation should be thread safe, but there may still be some loose ends there.

I'll look into adding more robust testing here, if there's any additional details about this particular incident, I'd love to know :) Also, when you say 10-100s of MBs for your GLTF, is that _all_ mesh data, or is that mostly textures?

**Reply by paulmelis** [reply]
Disclaimer: I've been testing on a Quest 1 through Link for now, will get back to the XR-3 later this week. I also updated to 0.3.6-preview4 to make sure I'm using the texture async loading. The latter I definitely see happening, but it's during loading of the mesh geometry itself that I'm seeing weird artifacts (parts of the view not updating, distortions). It happens both for meshes with lots of texture, but also for non-textured meshes. The biggest non-textured mesh I've tried is the Lucy PLY model, which is 28M triangles in a single mesh (.glb of around 600MB).

**Reply by paulmelis** [reply]
Just tried loading the lucy model on the XR-3 and it gives the same rendering lockups/artifacts with 0.3.6-preview4. After digging some more it seems that most of the issue was caused by my code to get the number of vertices/triangles, which you already explained can be done in a better way without a performance impact. 

However, not all of the artifacts go away if I disable that geometry stats code. I have a video clip of the headset view, in case you want to see yourself what's happening, but in short, parts of the 3D visuals are simply not drawn correctly anymore. Here's a screenshot:

![afbeelding](https://user-images.githubusercontent.com/1408287/160997908-50d5bd23-86f3-4ab4-abe8-7aef6a10d435.png)

I added printing of `Time.Totalf` and `Time.Elapsedf` at the beginning of the `while (SK.Step())` loop and the rendering loop stops for a few frames, apparently causing the image artifacts (as I assume the motion prediction still tries to do its job). The render loop lockup does appear to be due to the `Model.FromFile()` call on the separate loading thread, as the Model hasn't been passed to the main thread yet at that point:

```
[SK info] Time: 5.0846815, elapsed: 0.011172
[SK info] Time: 5.096025, elapsed: 0.0113434
[SK info] Time: 5.1071, elapsed: 0.011075
[SK info] Time: 5.118211, elapsed: 0.0111107
[SK info] Time: 5.1293883, elapsed: 0.0111777
[SK info] Time: 5.1403527, elapsed: 0.0109645
[SK info] Time: 5.151744, elapsed: 0.0113911
[SK info] Time: 5.163075, elapsed: 0.011331
[SK info] (model fetch thread) Reading file C:\Users\VR-1\Desktop\models\lucy2.glb (scene 'Lucy statue')
[SK info] (model fetch thread) calling Model.FromFile()
[SK info] Time: 5.1743693, elapsed: 0.0112943
[SK info] Time: 5.1851997, elapsed: 0.0108305
[SK info] Time: 5.196249, elapsed: 0.0110493
[SK info] Time: 5.2072315, elapsed: 0.0109826
[SK info] Time: 5.2179847, elapsed: 0.0107532
[SK info] Time: 5.2291446, elapsed: 0.0111596
[SK info] Time: 5.2403126, elapsed: 0.0111682
[SK info] Time: 5.251432, elapsed: 0.0111195
[SK info] Time: 5.2625594, elapsed: 0.0111273
[SK info] Time: 5.2736516, elapsed: 0.0110924
[SK info] Time: 5.285084, elapsed: 0.011432
[SK info] Time: 5.2965093, elapsed: 0.0114253
[SK info] Time: 5.3075566, elapsed: 0.0110476
[SK info] Time: 5.318908, elapsed: 0.0113515
[SK info] Time: 5.3298526, elapsed: 0.0109443
[SK info] Time: 5.340814, elapsed: 0.0109618
[SK info] Time: 5.351388, elapsed: 0.0105736
[SK info] Time: 5.362543, elapsed: 0.0111551
[SK info] Time: 5.3735805, elapsed: 0.0110373
[SK info] Time: 5.384584, elapsed: 0.0110035
[SK info] Time: 5.3956714, elapsed: 0.0110874
[SK info] Time: 5.4066987, elapsed: 0.0110276
[SK info] Time: 5.4178257, elapsed: 0.0111267
[SK info] Time: 5.428891, elapsed: 0.0110657
[SK info] Time: 5.4400864, elapsed: 0.0111951
[SK info] Time: 5.451492, elapsed: 0.0114056
[SK info] Time: 5.4627247, elapsed: 0.0112327
[SK info] Time: 5.4737916, elapsed: 0.0110669
[SK info] Time: 5.484682, elapsed: 0.0108906
[SK info] Time: 5.4960613, elapsed: 0.0113791
[SK info] Time: 5.507115, elapsed: 0.0110539
[SK info] Time: 5.518349, elapsed: 0.0112342
[SK info] Time: 5.5294566, elapsed: 0.0111072
[SK info] Time: 5.5402446, elapsed: 0.010788
[SK info] Time: 5.5512786, elapsed: 0.0110339
[SK info] Time: 5.562408, elapsed: 0.0111294
[SK info] Time: 5.5738354, elapsed: 0.0114274
[SK info] Time: 5.584909, elapsed: 0.0110739
[SK info] Time: 5.5961137, elapsed: 0.0112044
[SK info] Time: 5.6063027, elapsed: 0.0101893
# Here's the lockup
[SK info] Time: 6.153845, elapsed: 0.547542
[SK info] Time: 6.1619186, elapsed: 0.0080739
[SK info] Time: 6.17309, elapsed: 0.0111715
[SK info] (model fetch thread) Model.FromFile() returned
[SK info] Time: 6.185134, elapsed: 0.0120435
[SK info] Time: 6.1983943, elapsed: 0.0132607
[SK info] Have model update
[SK info] Time: 6.209526, elapsed: 0.0111318
[SK info] Time: 6.2213063, elapsed: 0.0117801
[SK info] Time: 6.2275934, elapsed: 0.0062873
[SK info] Time: 6.237293, elapsed: 0.009699
[SK info] Time: 6.2431216, elapsed: 0.0058289
[SK info] Time: 6.253124, elapsed: 0.0100026
[SK info] Time: 6.262891, elapsed: 0.0097666
```

Any suggestions on things I can try to nail down further what is going on?

**Reply by maluoi** [reply]
As far as rendering artifacts are concerned, I'd check to see what "extrapolation" looks like for Varjo's runtime! Every runtime has a slightly different visual response to the application dropping frames, most will reproject the existing content around until the app provides it with a new frame. On WMR, this can look a bit "bubbly" in extreme cases, and almost always, there's clipped content towards the edges. I suspect that's what's happening here, especially as you have identified a frame drop.

28M is a lot of triangles! Right now I do actually have the mesh data upload happening on the GPU thread, regardless of what thread you start things from, so that'd likely account for the .5s pause. I only really need that behavior for OpenGL though, I disabled this behavior for textures, but it looks like I forgot to disable that for meshes as well!
